shell/platform/windows/text_input_plugin.cc

```diff
@@ -0,0 +1,212 @@
+  }
+  // TODO(awdavies): Actually handle potential unicode characters. Probably
+  // requires some ICU data or something.
+  active_model_->AddCharacter(static_cast<char>(code_point));
+  SendStateUpdate(*active_model_);
+}
+
+void TextInputPlugin::KeyboardHook(FlutterWindow* window,
+                                   int key,
+                                   int scancode,
+                                   int action,
+                                   int mods) {
+  if (active_model_ == nullptr) {
+    return;
+  }
+  if (action == WM_KEYDOWN) {  // TODO repease
+    switch (key) {
+      case VK_LEFT:
+        if (active_model_->MoveCursorBack()) {
+          SendStateUpdate(*active_model_);
+        }
+        break;
+      case VK_RIGHT:
+        if (active_model_->MoveCursorForward()) {
+          SendStateUpdate(*active_model_);
+        }
+        break;
+      case VK_END:
+        active_model_->MoveCursorToEnd();
+        SendStateUpdate(*active_model_);
+        break;
+      case VK_HOME:
+        active_model_->MoveCursorToBeginning();
+        SendStateUpdate(*active_model_);
+        break;
+      case VK_BACK:
+        if (active_model_->Backspace()) {
+          SendStateUpdate(*active_model_);
+        }
+        break;
+      case VK_DELETE:
+        if (active_model_->Delete()) {
+          SendStateUpdate(*active_model_);
+        }
+        break;
+      case VK_RETURN:
+        EnterPressed(active_model_);
+        break;
+      default:
+        break;
+    }
+  }
+}
+
+TextInputPlugin::TextInputPlugin(flutter::BinaryMessenger* messenger)
+    : channel_(std::make_unique<flutter::MethodChannel<rapidjson::Document>>(
+          messenger,
+          kChannelName,
+          &flutter::JsonMethodCodec::GetInstance())),
+      active_model_(nullptr) {
+  channel_->SetMethodCallHandler(
+      [this](
+          const flutter::MethodCall<rapidjson::Document>& call,
+          std::unique_ptr<flutter::MethodResult<rapidjson::Document>> result) {
+        HandleMethodCall(call, std::move(result));
+      });
+}
+
+TextInputPlugin::~TextInputPlugin() = default;
+
+void TextInputPlugin::HandleMethodCall(
+    const flutter::MethodCall<rapidjson::Document>& method_call,
+    std::unique_ptr<flutter::MethodResult<rapidjson::Document>> result) {
+  const std::string& method = method_call.method_name();
+
+  if (method.compare(kShowMethod) == 0 || method.compare(kHideMethod) == 0) {
+    // These methods are no-ops.
+  } else if (method.compare(kClearClientMethod) == 0) {
+    active_model_ = nullptr;
+  } else {
+    // Every following method requires args.
+    if (!method_call.arguments() || method_call.arguments()->IsNull()) {
+      result->Error(kBadArgumentError, "Method invoked without args");
+      return;
+    }
+    const rapidjson::Document& args = *method_call.arguments();
+
+    if (method.compare(kSetClientMethod) == 0) {
+      // TODO(awdavies): There's quite a wealth of arguments supplied with this
```

Comment:
Fix TODO.

Reply:
It's fine to just remove it actually; we already have bugs tracking some things that will substantially alter this code, at which point this will be done as a side-effect.